### PR TITLE
3824 no smart quotes for nodes

### PR DESCRIPTION
**Pull request by tk0miya**
Refactor #3825 using `FixedTextElement`.
And I also added `addnodes.not_smartquotable` to represent the node does not support smart-quotes.

### CODE REVIEW

File: sphinx/addnodes.py [hunk 1]
```diff
@@ -91,7 +96,7 @@ class desc_signature(nodes.Part, nodes.Inline, nodes.TextElement):
     """
 
 
-class desc_signature_line(nodes.Part, nodes.Inline, nodes.TextElement):
+class desc_signature_line(nodes.Part, nodes.Inline, nodes.FixedTextElement):
```
Comment (jfbu) [inline sphinx/addnodes.py:99]: I think leaving TextElement should now be ok. Indeed the docutils.transforms.universal code contains:

```
# nested TextElements are not "block-level" elements:
            if isinstance(node.parent, nodes.TextElement):
                continue
```

So `desc_signature_line` is skipped as child of `desc_signature`. Only the latter is walked for its Text nodes. But now that you assign FixedTextElement class to `desc_signature` it is skipped earlier:
```
            # skip preformatted text blocks and special elements:
            if isinstance(node, (nodes.FixedTextElement, nodes.Special)):
                continue
```

**Review by jfbu — APPROVED** [review]
LGTM with nits.

- assigning class FixedTextElement to desc_signature for example means that children nodes of TextElement class need not be explicitely FixedTextElement for the smart quotes to be skipped (see comment)

- I have tested with keeping FixedTextElement only for production, desc_signature, parameter_list and parameter and it was enough in my unfortunately insufficient test file.

- what do you think of class Special ? it is also skipped by Docutils. Using FixedTextElement produces ``xml:space="preserve"`` in pseudoxml, I don't know if good or bad.

**Comment by jfbu** [comment]
In reviewing this I noticed two other problems.

First one is a latex one. With this

```
class Foo:
    """Docstring for 'my' class Foo."""

    #: Doc "comment" for class attribute Foo.bar.
    #: It can have multiple lines.
    bar = "1"

    flox = "1.5"   #: Doc "comment" for Foo.flox. One line only.

    baz = "2"
    """Docstring for class attribute Foo.baz."""

    def __init__(self):
        #: Doc comment for "instance" attribute qux.
        self.qux = "3"

        self.spam = "4"
        """Docstring for 'instance' attribute spam."""
```

the str variable values ``"3"`` etc...  are `desc_annotation` and end up as `\sphinxstrong{ = '3'}` for example. But the TeX mechanism transforms ``'`` in closing curly quote in PDF output. Earlier at 1.5.6, it would have been ``\sphinxstrong{ = `2'}`` giving curly quotes, not from smart quotes, but from TeX mechanism, specific to English. Thus I think earlier situation was buggy, but newer one is not much better. A Python `str` should be with straight quotes. This can be achieved if it ends up as `\sphinxcode` in tex file.

Second problem is #3877

**Comment by jfbu** [comment]
The FixedTextElement class is described as 

```
class FixedTextElement(TextElement):


    """An element which directly contains preformatted text."""
```

I don't know what "directly" means, but I see Docutils assigns the class to `address`, `literal_block`, `doctest_block`, `math_block`, `comment`. Perhaps they have property to contain text nodes "directly", i.e., without intermediate node types ? Is it then right to make says `desc_signature` inherit from FixedTextElement, as it does not in the document tree contain text nodes "directly", but contains them via intermediate nodes such as ``desc_annotation`` ?

In Docutils 0.14, there is 

```
    nodes_to_skip = (nodes.FixedTextElement, nodes.Special)
    """Do not apply "smartquotes" to instances of these block-level nodes."""
```

as class attribute of SmartQuotes class and its ``apply`` method obeys it. This would allow to simply add the class ``not_smartquotable``. But at Docutils 0.13 it is hard-coded in ``apply`` method of SmartQuotes class.

But as we support older Docutils, we would need to also override the ``apply`` method in ``SphinxSmartQuotes`` to get this working. So far we only modified ``get_tokens`` but if we need to override ``apply`` then letting ``SphinxSmartQuotes`` inherit from ``SmartQuotes`` is cosmetic, as it is entirely re-written...

**Comment by tk0miya** [comment]
Thank you for reviewing. As you said, `desc_signature ` should not inherit `FixedTextElement`.

**Comment by tk0miya** [comment]
Fixed.

>what do you think of class Special ?

I think `Special` is already skipped by `SmartQuotes.apply`. So I think we don't need to skip in our code.

>But as we support older Docutils, we would need to also override the apply method in SphinxSmartQuotes to get this working.

Really? I think overriding `get_tokens()` goes well. What case?

**Comment by jfbu** [comment]
The Docutils 0.13.1 has no hook into the ``apply()`` to extend the list of classes to skip beyond FixedTextElement and Special. Overriding `get_tokens()` allows only to influence the examination of text nodes, depending on their parents. Inside ``apply()`` there is earlier examination of TextElement classes which allows to skip children text nodes. And TextElement whose parent is TextElement is skipped automatically. So this is efficient way to speed-up tree traversal. Overriding `get_tokens()` is less expeditive and the check is only on parent of text node.

**Comment by tk0miya** [comment]
If your worry is only speed, we don't need to care that in my opinion. It means upgrading docutils is a best optimization for Sphinx.
In addition, `not_smartquotable` nodes are not too many in this moment (Now we are using only for `literal_emphasis` and `literal_strong`). So they might not affect to performance.

**Comment by tk0miya** [comment]
Now I merged this. Please rethink this again after you'll come back :-)

File: sphinx/transforms/__init__.py [hunk 2]
```diff
@@ -343,21 +343,5 @@ def get_tokens(self, txtnodes):
...
-                                            addnodes.desc_parameterlist,
-                                            addnodes.desc_signature_line,
-                                            addnodes.desc_type,
-                                            addnodes.production,
-                                            nodes.math,
-                                            nodes.image,
-                                            nodes.raw,
-                                            nodes.problematic))]
-            yield (nodetype, txtnode.astext())
+            smartquotable = not is_smartquotable(txtnode)
```
Comment (jfbu) [inline sphinx/transforms/__init__.py:346]: @tk0miya naming is confusing here, 

```
notsmartquotable = not is_smartquotable(txtnode)
yield (texttype[notsmartquotable], txtnode.astext())
```
would be clearer...

Comment (tk0miya) [reply sphinx/transforms/__init__.py:346]: Absolutely

Comment (jfbu) [reply sphinx/transforms/__init__.py:346]: Done at 0059c05a8

**Pull request merged**